_posts/2020-01-23-scatter-gather-pt2.md

```diff
@@ -0,0 +1,142 @@
+---
+layout: post
+title: Scatter-Gather in NServiceBus, part 2
+---
+
+In the [previous blog post](/2019/09/25/scatter-gather-pt1.html) we concluded that sagas struggle to deal with scenarios where a multiple messages need to update the same saga instance. Depending on the persistence option, this can manifest differently in practice:
+* For persistence options using pessimistic locking, the described scneario will slow down the endpoint's throughput as the locking mechanism will only allow a single message to be processed at a time per saga instance.
+* For persistence options using optimistic concurrency control, the endpoint can run into many concurrency violation exceptions. These exceptions will force the losing transactions to retry, potentially causing large amounts of retries. At some point, messages might even be moved to the error queue due to exceding the configured retry limits.
+
+If you'd have to chose between one of the options, you'd probably prefer the pessimistic locking approach. The throughput limitations are probably less severe than the impact of multiple retries and are way better than dealing with messages in the error queue. Luckily, most of the supported Saga persistence options support pessimistic locking for sagas by now (the MongoDB and ServiceFabric persistence packages just recently switched to pessimistic locking). Particular Software also vastly improved the official [documentation on saga concurrency](https://docs.particular.net/nservicebus/sagas/concurrency), describing different concurrency problems and approaches to solve those, a highly recommended read.
+
+In most cases, using a persistence option with pessimistic locking capabilities will prevent major problems for you, at the cost of slower throughput at some times. However, there is another approach which doesn't impact the endpoint throughput and also won't cause concurrency exceptions and retries. It's a lot more complex than just using pessimistic locking but sometimes the throughput limitations can be an important factor or you might rely on a persistence not support pessimistic locking. Here's a rough overview how this approach will work:
+
+1. Instead of aggregating the results in the saga directly, a simple message handler will receive the response message type.
+2. The handler will store the response data in an append-only storage which allows extremely fast inserts without concurrency violations.
+3. A separate process will repeatedly check if all expected responses arrived and will continue the business process managed by the saga.
```

Comment:
```suggestion
3. A separate process repeatedly checks if all expected responses arrived and will continue the business process managed by the saga.
```

I'd change the word "process" here with something else. I immediately connected it to a process running on a machine, rather than a generic business process. In the end the thing checking could be the saga itself using a timeout.